### ISSUE TITLE
Linux Build Warnings

### ISSUE THREAD

**Issue by ptheywood** [FLAMEGPU/FLAMEGPU#31]
Warnings are generated during compilation under linux, which should be resolved where possible.

```
In file included from src/visualisation/CustomVisualisation.cpp:26:0:
src/visualisation/CustomVisualisation.h:50:8: warning: extra tokens at end of #endif directive
 #endif __VISUALISATION
        ^
In file included from src/visualisation/CustomVisualisation.cpp:27:0:
src/visualisation/GLUTInputController.h:55:8: warning: extra tokens at end of #endif directive
 #endif __GLUT_INPUT_CONTROLLER
```

...

```
 src/visualisation/CustomVisualisation.cpp: In function ‘void initVisualisation()’:
src/visualisation/CustomVisualisation.cpp:65:45: warning: deprecated conversion from string constant to ‘char*’ [-Wwrite-strings]
     char *argv[] = {"GLUT application", NULL}; 
```

...

```
src/visualisation/MenuDisplay.cpp: In function ‘void drawInfoDisplay(int, int)’:
src/visualisation/MenuDisplay.cpp:206:63: warning: deprecated conversion from string constant to ‘char*’ [-Wwrite-strings]
   printInfoLine("********** Simulation Information **********");
```

**Comment by mondus** [comment]
This is annoying. Should use the following fix in all places.

`#endif //__VISUALISATION`

It is good practice to label the `endif` macros if they span large numbers of lines.

**Comment by ptheywood** [comment]
> Any functions into which you pass string literals "I am a string literal" should use char const * as the type instead of char*.

https://stackoverflow.com/questions/59670/how-to-get-rid-of-deprecated-conversion-from-string-constant-to-char-warnin

` #endif __VISUALISATION` to ` #endif // __VISUALISATION`

Might be dependant on gcc version

**Comment by ptheywood** [comment]
I'm getting these warnings on `Ubuntu 16.10`, with `g++ 4.9.4` for the Sugarscape example.

**Comment by ptheywood** [comment]
With `-Wall`, i've spotted a set of extra warnings reported by `gcc` / `g++`. 

```
src/dynamic/simulation.cu:20:0: warning: ignoring #pragma warning  [-Wunknown-pragmas]
 #pragma warning(push)
 ^
src/dynamic/simulation.cu:21:0: warning: ignoring #pragma warning  [-Wunknown-pragmas]
 #pragma warning (disable : 4267)
 ^
src/dynamic/simulation.cu:22:0: warning: ignoring #pragma warning  [-Wunknown-pragmas]
 #pragma warning (disable : 4244)
 ^
src/dynamic/simulation.cu:42:0: warning: ignoring #pragma warning  [-Wunknown-pragmas]
 #pragma warning(pop)
 ^
```

caused by the following code in `simulation.xslt`

```
#pragma warning(push)
#pragma warning (disable : 4267)
#pragma warning (disable : 4244)
```

These pragmas do not directly effect the code, but suppress the related warnigs in `msvc`.
We should probably add the `gcc` equivalents and wrap the appropriate pragmas in `ifdefs` which capture the appropriate compiler.

**Comment by ptheywood** [comment]
a842d10 resolves the #pragma issue, this will be resolved when #108 is merged

**Comment by ptheywood** [comment]
when specifying higher compilation levels on windows (`/W4` or `/Wall`) additional warnings are produced.

`Wall` on windows produces hundreds of (in my opinion, and others) pointless warnings, so `W4` is the saner alternative for high levels of warnings.

`W4` does produce a much saner set of warnings, but still includes some excessive warnings, such as the following:

```
# many thrust / cub warnings
# ...
c:\users\ptheywood\code\flamegpu\examples\gameoflife\src\dynamic\header.h(91): warning C4324: 'xmachine_memory_cell': structure was padded due to alignment specifier
c:\users\ptheywood\code\flamegpu\examples\gameoflife\src\dynamic\FLAMEGPU_kernals.cu(557): warning C4505: 'RNG_rand48_iterate_single': unreferenced local function has been removed
src/dynamic/main.cu(39): warning C4100: 'argv': unreferenced formal parameter
```
